Implement parsing/serialization for text-emphasis-position

https://developer.mozilla.org/en-US/docs/Web/CSS/text-emphasis-position

The grammar is `[ over | under ] && [ right | left ]`, which means that must have one of each keyword from both sets (`over`/`under` and `right`/`left`). We can also view this as `[over | under] [right | left] | [right | left] [over | under]`, i.e. it is "'`over` or `under`' followed by '`right` or `left`', OR '`right` or `left`' followed by '`over` or `under`'"

See https://github.com/servo/servo/wiki/Property-hacking-guide for instructions on how to work on such bugs. This bug requires familiarity with Rust.

File: https://github.com/servo/servo/blob/master/components/style/properties/longhand/inherited_text.mako.rs

@Manishearth I have the code for parsing here: https://github.com/AgostonSzepessy/servo/blob/text-emphasis-position/components/style/properties/longhand/inherited_text.mako.rs#L939-L1049 but I get a compilation error that I don't know how to resolve:

```
error[E0407]: method `Text_emphasis_position` is not a member of trait `CSSStyleDeclarationMethods`
   --> /home/agoston/Documents/stuff/servo/components/script/dom/cssstyledeclaration.rs:42:13
    |
42  |             fn $getter(&self) -> DOMString {
    |             ^ not a member of trait `CSSStyleDeclarationMethods`
...
374 |     css_properties_accessors!(css_properties);
    |                               - in this macro invocation

error[E0407]: method `SetText_emphasis_position` is not a member of trait `CSSStyleDeclarationMethods`
   --> /home/agoston/Documents/stuff/servo/components/script/dom/cssstyledeclaration.rs:45:13
    |
45  |             fn $setter(&self, value: DOMString) -> ErrorResult {
    |             ^ not a member of trait `CSSStyleDeclarationMethods`
...
374 |     css_properties_accessors!(css_properties);
    |                               - in this macro invocation

error[E0407]: method `TextEmphasisPosition` is not a member of trait `CSSStyleDeclarationMethods`
   --> /home/agoston/Documents/stuff/servo/components/script/dom/cssstyledeclaration.rs:42:13
    |
42  |             fn $getter(&self) -> DOMString {
    |             ^ not a member of trait `CSSStyleDeclarationMethods`
...
374 |     css_properties_accessors!(css_properties);
    |                               - in this macro invocation

error[E0407]: method `SetTextEmphasisPosition` is not a member of trait `CSSStyleDeclarationMethods`
   --> /home/agoston/Documents/stuff/servo/components/script/dom/cssstyledeclaration.rs:45:13
    |
45  |             fn $setter(&self, value: DOMString) -> ErrorResult {
    |             ^ not a member of trait `CSSStyleDeclarationMethods`
...
374 |     css_properties_accessors!(css_properties);
    |                               - in this macro invocation

error: aborting due to 4 previous errors

error: Could not compile `script`.
```

How do I fix this error?

--- a/components/style/properties/longhand/inherited_text.mako.rs
+++ b/components/style/properties/longhand/inherited_text.mako.rs
@@ -936,6 +936,55 @@ ${helpers.single_keyword("text-align-last",
     }
 </%helpers:longhand>
 
+<%helpers:longhand name="text-emphasis-position" animatable="False" products="none">
+    use std::fmt;
+    use values::computed::ComputedValueAsSpecified;
+    use values::NoViewportPercentage;
+    use style_traits::ToCss;
+
+    define_css_keyword_enum!(HorizontalWritingModeValue:
+                             "over" => Over,
+                             "under" => Under);
+    define_css_keyword_enum!(VerticalWritingModeValue:
+                             "right" => Right,
+                             "left" => Left);
+
+    #[derive(Debug, Clone, PartialEq)]
+    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    pub struct SpecifiedValue(pub HorizontalWritingModeValue, pub VerticalWritingModeValue);
+
+    pub mod computed_value {
+        pub type T = super::SpecifiedValue;
+    }
+
+    impl ComputedValueAsSpecified for SpecifiedValue {}
+    impl NoViewportPercentage for SpecifiedValue {}
+
+    pub fn get_initial_value() -> computed_value::T {
+        SpecifiedValue(HorizontalWritingModeValue::Over, VerticalWritingModeValue::Right)
+    }
+
+    pub fn parse(_context: &ParserContext, input: &mut Parser) -> Result<SpecifiedValue, ()> {
+       if let Ok(horizontal) = input.try(|input| HorizontalWritingModeValue::parse(input)) {
+            let vertical = try!(VerticalWritingModeValue::parse(input));
+            Ok(SpecifiedValue(horizontal, vertical))
+        } else {
+            let vertical = try!(VerticalWritingModeValue::parse(input));
+            let horizontal = try!(HorizontalWritingModeValue::parse(input));
+            Ok(SpecifiedValue(horizontal, vertical))
+        }
+    }
+
+    impl ToCss for SpecifiedValue {
+        fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+            let SpecifiedValue(horizontal, vertical) = *self;
+            try!(horizontal.to_css(dest));
+            try!(dest.write_char(' '));
+            vertical.to_css(dest)
+        }
+    }
+</%helpers:longhand>
+
 // TODO(pcwalton): `full-width`
 ${helpers.single_keyword("text-transform",
                          "none capitalize uppercase lowercase",

--- a/tests/unit/style/parsing/inherited_text.rs
+++ b/tests/unit/style/parsing/inherited_text.rs
@@ -48,3 +48,34 @@ fn text_emphasis_style_longhand_should_parse_properly() {
     let devanagari_string_struct = SpecifiedValue::String("षि".to_string());
     assert_eq!(devanagari_string, devanagari_string_struct);
 }
+
+#[test]
+fn test_text_emphasis_position() {
+    use style::properties::longhands::text_emphasis_position;
+    use style::properties::longhands::text_emphasis_position::{HorizontalWritingModeValue, VerticalWritingModeValue };
+    use style::properties::longhands::text_emphasis_position::SpecifiedValue;
+
+    let over_right = parse_longhand!(text_emphasis_position, "over right");
+    assert_eq!(over_right, SpecifiedValue(HorizontalWritingModeValue::Over, VerticalWritingModeValue::Right));
+
+    let over_left = parse_longhand!(text_emphasis_position, "over left");
+    assert_eq!(over_left, SpecifiedValue(HorizontalWritingModeValue::Over, VerticalWritingModeValue::Left));
+
+    let under_right = parse_longhand!(text_emphasis_position, "under right");
+    assert_eq!(under_right, SpecifiedValue(HorizontalWritingModeValue::Under, VerticalWritingModeValue::Right));
+
+    let under_left = parse_longhand!(text_emphasis_position, "under left");
+    assert_eq!(under_left, SpecifiedValue(HorizontalWritingModeValue::Under, VerticalWritingModeValue::Left));
+
+    let right_over = parse_longhand!(text_emphasis_position, "right over");
+    assert_eq!(right_over, SpecifiedValue(HorizontalWritingModeValue::Over, VerticalWritingModeValue::Right));
+
+    let left_over = parse_longhand!(text_emphasis_position, "left over");
+    assert_eq!(left_over, SpecifiedValue(HorizontalWritingModeValue::Over, VerticalWritingModeValue::Left));
+
+    let right_under = parse_longhand!(text_emphasis_position, "right under");
+    assert_eq!(right_under, SpecifiedValue(HorizontalWritingModeValue::Under, VerticalWritingModeValue::Right));
+
+    let left_under = parse_longhand!(text_emphasis_position, "left under");
+    assert_eq!(left_under, SpecifiedValue(HorizontalWritingModeValue::Under, VerticalWritingModeValue::Left));
+}

